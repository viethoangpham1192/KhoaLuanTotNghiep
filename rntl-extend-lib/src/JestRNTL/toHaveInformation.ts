import type { ReactTestInstance } from "react-test-renderer";
import {
  getAllPlaceholderOf,
  getAllTextOf,
  getAllTypeOf,
  getAllValueOf,
} from "../GetInformation/getAllInformation";
import { receivedToIncludeThreshold } from "../utils";
import { getParentOf, getPropsOf } from "../GetInformation";

function toHaveText(received: ReactTestInstance, threshold: string | string[]) {
  const receivedArray: string[] = getAllTextOf(received);
  let pass: boolean = false;

  if (Array.isArray(threshold)) {
    pass = receivedToIncludeThreshold(receivedArray, threshold);
  } else {
    pass = receivedToIncludeThreshold(receivedArray, [threshold]);
  }

  if (pass) {
    return {
      message: () => `Expected ${received} to have text ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () => `Expected ${received} not to have text ${threshold}`,
      pass: false,
    };
  }
}

function toHavePlaceholder(
  received: ReactTestInstance,
  threshold: string | string[]
) {
  const receivedArray: string[] = getAllPlaceholderOf(received);
  let pass: boolean = false;

  if (Array.isArray(threshold)) {
    pass = receivedToIncludeThreshold(receivedArray, threshold);
  } else {
    pass = receivedToIncludeThreshold(receivedArray, [threshold]);
  }

  if (pass) {
    return {
      message: () => `Expected ${received} to have placeholder ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () =>
        `Expected ${received} not to have placeholder ${threshold}`,
      pass: false,
    };
  }
}

function toHaveType(received: ReactTestInstance, threshold: string | string[]) {
  const receivedArray: string[] = getAllTypeOf(received);
  let pass: boolean = false;

  if (Array.isArray(threshold)) {
    pass = receivedToIncludeThreshold(receivedArray, threshold);
  } else {
    pass = receivedToIncludeThreshold(receivedArray, [threshold]);
  }

  if (pass) {
    return {
      message: () => `Expected ${received} to have type ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () => `Expected ${received} not to have type ${threshold}`,
      pass: false,
    };
  }
}

function toIncludeComponent(
  this: jest.MatcherContext,
  received: ReactTestInstance,
  threshold: ReactTestInstance
) {
  let pass: boolean = false;
  let parent: ReactTestInstance | null = getParentOf(threshold);

  while (parent) {
    if (this.equals(received, parent)) {
      pass = true;
      parent = null;
    } else {
      parent = getParentOf(parent);
    }
  }

  if (pass) {
    return {
      message: () => `Expected ${received} to be parent of ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () => `Expected ${received} to be not parent of ${threshold}`,
      pass: false,
    };
  }
}

function toHaveProps(
  this: jest.MatcherContext,
  received: ReactTestInstance,
  threshold: object
) {
  let pass: boolean = false;
  const obj1: Record<string, any> = getPropsOf(received);
  let obj: Record<string, any> = { ...threshold };

  for (let key in threshold) {
    if (this.equals(obj1[key], obj[key])) {
      delete obj[key];
    }
  }

  pass = Object.keys(obj).length === 0;

  if (pass) {
    return {
      message: () => `Expected ${received} to have prop ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () => `Expected ${received} to be not have props ${threshold}`,
      pass: false,
    };
  }
}

function toHaveValue(
  this: jest.MatcherContext,
  received: ReactTestInstance,
  threshold: (string | boolean | number)[] | string | boolean | number
) {
  let pass: boolean = false;
  if (typeof threshold === "object") {
    const r = JSON.stringify(getAllValueOf(received));
    let arr = [...threshold];
    for (let i = arr.length - 1; i >= 0; i--) {
      if (r.includes(JSON.stringify(arr[i]))) {
        arr.splice(i, 1);
      }
    }
    pass = arr.length === 0;
  } else {
    const r = JSON.stringify(getAllValueOf(received));
    pass = r.includes(threshold.toString());
  }

  if (pass) {
    return {
      message: () => `Expected ${received} to have value ${threshold}`,
      pass: true,
    };
  } else {
    return {
      message: () => `Expected ${received} to be not have value ${threshold}`,
      pass: false,
    };
  }
}

export {
  toHaveText,
  toHavePlaceholder,
  toHaveType,
  toIncludeComponent,
  toHaveProps,
  toHaveValue,
};
