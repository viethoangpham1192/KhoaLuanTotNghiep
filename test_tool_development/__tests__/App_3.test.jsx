import {fireEvent, render, screen} from '@testing-library/react-native';
import Login from '../src/components/Login/Login';
import {getAllParentOf} from '../src/test-tool/getAllInformation';
import {getChildrenOf, getValueOf} from '../src/test-tool/getInformation';
import {toHaveValue} from '../src/test-tool/toHaveInformation';

test('Test snapshot', async () => {
  const {toJSON} = render(<Login title="hell" />);
  console.log(toJSON());

  expect(toJSON()).toMatchSnapshot();
});

expect.extend({
  toHaveValue,
});
