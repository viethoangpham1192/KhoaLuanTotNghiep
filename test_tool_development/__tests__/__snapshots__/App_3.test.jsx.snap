// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot 1`] = `
<View
  testID="Component"
>
  <Text
    testID="title"
  >
    hello
  </Text>
  <Text>
    Account
  </Text>
  <TextInput
    defaultValue="Account"
    onChangeText={[Function]}
    placeholder="Account"
    style={
      {
        "color": "yellow",
      }
    }
    value="Account"
  />
  <Text>
    Password
  </Text>
  <TextInput
    defaultValue="Password"
    onChangeText={[Function]}
    placeholder="Password"
    value="Password"
  />
  <RCTSwitch
    accessibilityRole="switch"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onTintColor="#81b0ff"
    style={
      [
        {
          "height": 31,
          "width": 51,
        },
        [
          {
            "color": "white",
          },
          {
            "backgroundColor": "#3e3e3e",
            "borderRadius": 16,
          },
        ],
      ]
    }
    testID="switch"
    thumbTintColor="#f5dd4b"
    tintColor="#767577"
    value={false}
  />
  <View>
    <View
      accessibilityLabel="Learn more about this purple button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="RealButton"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              {
                "color": "#841584",
              },
            ]
          }
        >
          Learn More
        </Text>
      </View>
    </View>
  </View>
</View>
`;
